File: phase-2 -full stack todo app/backend/src/api/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from sqlmodel import Session
from ..database.database import get_sync_session
from ..models.task_model import Task, TaskCreate, TaskRead, TaskUpdate, TaskUpdateCompletion
from ..models.user_model import User
from ..services.task_service import TaskService
from ..services.auth_service import AuthService
from .auth import get_current_user, get_user_id_from_header
from uuid import UUID

router = APIRouter()

@router.get("/tasks", response_model=List[TaskRead])
def get_tasks(
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Get all tasks for the specified user
    """
    # Validate that the user exists
    user_exists = AuthService.validate_user_exists(session, user_id)
    if not user_exists:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )

    tasks = TaskService.get_tasks_by_user_id(session, user_id)
    return tasks


@router.post("/tasks", response_model=TaskRead, status_code=status.HTTP_201_CREATED)
def create_task(
    task_data: TaskCreate,
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Create a new task for the specified user
    """
    # Validate that the user exists
    user_exists = AuthService.validate_user_exists(session, user_id)
    if not user_exists:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )

    # Override the user_id in the task data with the authenticated user's ID
    # This prevents users from creating tasks for other users
    task_data_dict = task_data.model_dump()
    task_data_dict['user_id'] = user_id
    validated_task_data = TaskCreate(**task_data_dict)

    task = TaskService.create_task(session, validated_task_data)
    return task


@router.get("/tasks/{task_id}", response_model=TaskRead)
def get_task(
    task_id: str,
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Get a specific task by ID for the specified user
    """
    task = TaskService.get_task_by_id_and_user_id(session, task_id, user_id)

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found or does not belong to user"
        )

    return task


@router.put("/tasks/{task_id}", response_model=TaskRead)
def update_task(
    task_id: str,
    task_update: TaskUpdate,
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Update a specific task by ID for the specified user
    """
    updated_task = TaskService.update_task(session, task_id, user_id, task_update)

    if not updated_task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found or does not belong to user"
        )

    return updated_task


@router.delete("/tasks/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(
    task_id: str,
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Delete a specific task by ID for the specified user
    """
    success = TaskService.delete_task(session, task_id, user_id)

    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found or does not belong to user"
        )

    return


@router.patch("/tasks/{task_id}/complete", response_model=TaskRead)
def update_task_completion(
    task_id: str,
    completion_update: TaskUpdateCompletion,
    user_id: str = Depends(get_user_id_from_header),
    session: Session = Depends(get_sync_session)
):
    """
    Update the completion status of a specific task by ID for the specified user
    """
    updated_task = TaskService.update_task_completion(session, task_id, user_id, completion_update)

    if not updated_task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found or does not belong to user"
        )

    return updated_task

File: phase-2 -full stack todo app/frontend/src/services/tasks.ts
import api from './api';
import { Task } from '@/types/task';
import logger from '../utils/logger';

class TaskService {
  /**
   * Get all tasks for the authenticated user
   */
  async getTasks(): Promise<Task[]> {
    try {
      logger.info('Fetching all tasks for authenticated user');
      const response = await api.get('/api/tasks');
      logger.info('Tasks fetched successfully', { count: response.data.tasks?.length || response.data.length });
      return response.data.tasks || response.data;
    } catch (error: any) {
      logger.error('Error fetching tasks', { error: error.response?.data?.message || error.message }, error);
      throw new Error(error.response?.data?.message || 'Failed to fetch tasks');
    }
  }

  /**
   * Create a new task
   */
  async createTask(taskData: Omit<Task, 'id' | 'created_at' | 'updated_at' | 'user_id'>): Promise<Task> {
    try {
      logger.info('Creating new task', { title: taskData.title });
      const response = await api.post('/api/tasks', taskData);
      logger.info('Task created successfully', { taskId: response.data.id, title: response.data.title });
      return response.data;
    } catch (error: any) {
      logger.error('Error creating task', {
        title: taskData.title,
        error: error.response?.data?.message || error.message
      }, error);
      throw new Error(error.response?.data?.message || 'Failed to create task');
    }
  }

  /**
   * Update an existing task
   */
  async updateTask(id: string, updates: Partial<Task>): Promise<Task> {
    try {
      logger.info('Updating task', { taskId: id, updates: Object.keys(updates) });
      const response = await api.put(`/api/tasks/${id}`, updates);
      logger.info('Task updated successfully', { taskId: response.data.id, title: response.data.title });
      return response.data;
    } catch (error: any) {
      logger.error('Error updating task', {
        taskId: id,
        updates,
        error: error.response?.data?.message || error.message
      }, error);
      throw new Error(error.response?.data?.message || 'Failed to update task');
    }
  }

  /**
   * Delete a task
   */
  async deleteTask(id: string): Promise<void> {
    try {
      logger.info('Deleting task', { taskId: id });
      await api.delete(`/api/tasks/${id}`);
      logger.info('Task deleted successfully', { taskId: id });
    } catch (error: any) {
      logger.error('Error deleting task', {
        taskId: id,
        error: error.response?.data?.message || error.message
      }, error);
      throw new Error(error.response?.data?.message || 'Failed to delete task');
    }
  }

  /**
   * Toggle task completion status
   */
  async toggleTaskCompletion(id: string): Promise<Task> {
    try {
      logger.info('Toggling task completion', { taskId: id });
      const response = await api.patch(`/api/tasks/${id}/complete`, { completed: true });
      logger.info('Task completion toggled successfully', {
        taskId: response.data.id,
        completed: response.data.completed
      });
      return response.data;
    } catch (error: any) {
      logger.error('Error toggling task completion', {
        taskId: id,
        error: error.response?.data?.message || error.message
      }, error);
      throw new Error(error.response?.data?.message || 'Failed to toggle task completion');
    }
  }

  /**
   * Get a specific task by ID
   */
  async getTaskById(id: string): Promise<Task> {
    try {
      logger.info('Fetching specific task', { taskId: id });
      const response = await api.get(`/api/tasks/${id}`);
      logger.info('Task fetched successfully', { taskId: response.data.id, title: response.data.title });
      return response.data;
    } catch (error: any) {
      logger.error('Error fetching task by ID', {
        taskId: id,
        error: error.response?.data?.message || error.message
      }, error);
      throw new Error(error.response?.data?.message || 'Failed to fetch task');
    }
  }
}

export default new TaskService();